go/arrow/compute/internal/kernels/vector_run_end_encode.go

```diff
@@ -0,0 +1,947 @@
+			inputOffset:   input.Offset,
+			inputValidity: input.Buffers[0].Buf,
+			inputValues:   input.Buffers[1].Buf,
+			valueType:     input.Type,
+			width:         input.Type.(*arrow.FixedSizeBinaryType).ByteWidth,
+		}
+	}
+	return nil
+}
+
+type encoder interface {
+	CountNumberOfRuns() (numValid, numOutput int64)
+	PreallocOutput(*exec.KernelCtx, int64, *exec.ExecResult)
+	WriteEncodedRuns(*exec.ExecResult) int64
+}
+
+func runEndEncodeImpl[R RunEndsType](ctx *exec.KernelCtx, batch *exec.ExecSpan, out *exec.ExecResult) error {
+	// first pass: count the number of runs
+	var (
+		inputArr      = &batch.Values[0].Array
+		inputLen      = inputArr.Len
+		numOutputRuns int64
+		numValidRuns  int64
+		enc           encoder
+	)
+
+	if inputLen == 0 {
+		reeType := arrow.RunEndEncodedOf(exec.GetDataType[R](), inputArr.Type)
+		*out = exec.ExecResult{
+			Type: reeType,
+			Children: []exec.ArraySpan{
+				{Type: reeType.RunEnds()}, {Type: reeType.Encoded()},
+			},
+		}
+		return nil
+	}
+
+	if err := validateRunEndType[R](inputLen); err != nil {
+		return err
+	}
+
+	enc = newEncoder[R](inputArr)
+	numValidRuns, numOutputRuns = enc.CountNumberOfRuns()
+	enc.PreallocOutput(ctx, numOutputRuns, out)
+
+	out.Children[1].Nulls = numOutputRuns - numValidRuns
+
+	written := enc.WriteEncodedRuns(out)
+	debug.Assert(written == numOutputRuns, "mismatch number of written values")
+	return nil
+}
+
+func runEndEncodeExec(ctx *exec.KernelCtx, batch *exec.ExecSpan, out *exec.ExecResult) error {
+	reeType := ctx.State.(RunEndEncodeState).RunEndType
+	switch reeType.ID() {
+	case arrow.INT16:
+		return runEndEncodeImpl[int16](ctx, batch, out)
+	case arrow.INT32:
+		return runEndEncodeImpl[int32](ctx, batch, out)
+	case arrow.INT64:
+		return runEndEncodeImpl[int64](ctx, batch, out)
+	}
+
+	return fmt.Errorf("%w: bad run end type %s", arrow.ErrInvalid, reeType)
+}
+
+type decodeBool[R RunEndsType] struct {
+	inputLen, inputOffset int64
+	inputRunEnds          []R
+
+	inputPhysicalOffset int64
+	inputValidity       []byte
+	inputValues         []byte
+	inputValueOffset    int64
+}
+
+func (de *decodeBool[R]) PreallocOutput(ctx *exec.KernelCtx, out *exec.ExecResult) {
+	*out = exec.ExecResult{
+		Type: arrow.FixedWidthTypes.Boolean,
+		Len:  de.inputLen,
+	}
+
+	if len(de.inputValidity) != 0 {
+		out.Buffers[0].WrapBuffer(ctx.AllocateBitmap(de.inputLen))
+	}
+
+	out.Buffers[1].WrapBuffer(ctx.AllocateBitmap(de.inputLen))
+}
+
+func (de *decodeBool[R]) ExpandAllRuns(out *exec.ExecResult) int64 {
+	var (
+		writeOffset         int64
+		runLength, numValid int64
+		outputValues        = out.Buffers[1].Buf
+		prevRunEnd          = R(de.inputOffset)
+	)
+
+	for i, runEnd := range de.inputRunEnds[de.inputPhysicalOffset:] {
+		runLength, prevRunEnd = int64(runEnd-prevRunEnd), runEnd
+		// if this run is a null, clear the bits and update writeOffset
+		if len(de.inputValidity) != 0 && bitutil.BitIsNotSet(de.inputValidity, int(de.inputValueOffset+de.inputPhysicalOffset)+i) {
```

Comment:
`inputValueOffset` could be just `inputOffset` here, right? It refers to the logical offset of the parent array. (?)

Reply:
* `de.inputOffset` is the logical offset of the parent array
* `de.inputPhysicalOffset` is the physical offset into the values and runs children arrays
* `de.inputValueOffset` is the internal offset of the values child array

For example, let's say you have the array `[1, 2, 3, 4, 5]`. Then let's take a slice of that, `[2, 3, 4]` and use that as our `values` for a run-end-encoded array. Finally let's add a logical offset to the REE array.

To make up some runs, that gives us the final REE array of:

* Parent: logical offset = 20, len = 30
    * Runs: offset = 0, len = 3, `[15, 30, 50]`
    * Values: offset = 1, len = 3, `[1, 2, 3, 4, 5]`

To get the proper values for the runs, we first find the physical offset (`1`) from the logical offset. Then when we index into the `values` child, we have to also apply the child's offset (`1`). Giving us the first value of `3` (index = `2`). This of course also applies to determining the bitmap validity for nulls.

Does that make sense? This is also tested in the unit tests.

Reply:
Oh right. I encapsulated much of this in a forward iterator in the C++ version and had forgotten there are 3 offset variables to address a physical value sometimes :)

Reply:
yea, Go doesn't really have `iterators` so there's just the `merged runs` iterator thing, but that's designed for iterating across 2 REE arrays simultaneously and getting the runs that line up rather than just iterating a single REE array. Maybe one day Go will provide a way to hook into the `for ... := range` construction for custom types lol

Reply:
> so there's just the merged runs iterator thing, but that's designed for iterating across 2 REE arrays simultaneously

I took that thigk apart in the C++ version and made a similar class that takes a single `ArraySpan` and allows iterating on the runs. It's not completely C++ iterator compatible, so I think a similar thing could be achieved with Go. Perhaps Go doesn't inline as aggressively and it wouldn't be as efficient.